dgc/Sign1CoseMessage.cs

```diff
@@ -0,0 +1,128 @@
+
+            var coseMsg = new Sign1CoseMessage();
+            coseMsg.Content = cborMsg[2].GetByteString();
+            coseMsg.Signature = cborMsg[3].GetByteString();
+
+            var algKey = protectedMap[HeaderKey_Alg];
+            if (algKey.AsInt32() == Alg_ES256.AsInt32())
+            {
+                coseMsg.RegisteredAlgorithm = HCertSupportedAlgorithm.ES256;
+            }
+            else if (algKey.AsInt32() == Alg_PS256.AsInt32())
+            {
+                coseMsg.RegisteredAlgorithm = HCertSupportedAlgorithm.PS256;
+            }
+            else
+            {
+                throw new NotSupportedException("Algorithm not supported");
+            }
+
+            if (unprotectedMap[HeaderKey_KID] != null)
+            {
+                var kidBytes = unprotectedMap[HeaderKey_KID].GetByteString();
+                coseMsg.KID = Encoding.UTF8.GetString(kidBytes);
+            }
+            else
+            {
+                var kidBytes = protectedMap[HeaderKey_KID].GetByteString();
+                coseMsg.KID = Encoding.UTF8.GetString(kidBytes);
+            }
+            return coseMsg;
+        }
+
+        public byte[] EncodeToBytes()
+        {
+            var coseObj = CBORObject.NewArray();
+
+            var protectedMap = CBORObject.NewMap();
+            if (RegisteredAlgorithm == HCertSupportedAlgorithm.ES256)
+            {
+                protectedMap[HeaderKey_Alg] = Alg_ES256;
+            }
+            else if (RegisteredAlgorithm == HCertSupportedAlgorithm.PS256)
+            {
+                protectedMap[HeaderKey_Alg] = Alg_PS256;
+            }
+
+            coseObj.Add(protectedMap.EncodeToBytes());
+
+            var unProtectedMap = CBORObject.NewMap();
+            var kidBytes = Encoding.UTF8.GetBytes(KID);
+            unProtectedMap[HeaderKey_KID] = CBORObject.FromObject(kidBytes);
+            coseObj.Add(unProtectedMap);
+
+            coseObj.Add(CBORObject.FromObjectAndTag(Content, 2));
+            coseObj.Add(CBORObject.FromObjectAndTag(Signature, 3));
+            return CBORObject.FromObjectAndTag(coseObj, Sign1Tag).EncodeToBytes();
+        }
+
+        public void Sign(AsymmetricCipherKeyPair keypair, HCertSupportedAlgorithm alg)
+        {
+            ISigner signer;
+            if (alg == HCertSupportedAlgorithm.ES256)
+            {
+                RegisteredAlgorithm = HCertSupportedAlgorithm.ES256;
+
+                signer = SignerUtilities.GetSigner("SHA-256withECDSA");
+            }
+            else if (alg == HCertSupportedAlgorithm.PS256)
+            {
+                RegisteredAlgorithm = HCertSupportedAlgorithm.PS256;
+
+                signer = SignerUtilities.GetSigner("SHA256withRSA/PSS");
+            }
+            else
+            {
+                throw new NotSupportedException("Algorithm not supported");
+            }
+
+            signer.Init(true, keypair.Private);
+            signer.BlockUpdate(Content, 0, Content.Length);
```

Comment:
Are you sure that you can sign only the content, because according to RFC 8152

https://tools.ietf.org/html/rfc8152#section-4.4

"In order to create a signature, a well-defined byte stream is needed.
   The Sig_structure is used to create the canonical form."

And the form is:

Sig_structure = [
       context : "Signature" / "Signature1" / "CounterSignature",
       body_protected : empty_or_serialized_map,
       ? sign_protected : empty_or_serialized_map,
       external_aad : bstr,
       payload : bstr
   ]

Reply:
Hi!
You are correct.
I will update this, I haven't been working on this. Been waiting for them to finalize the schema that I wanted to include in this and do way more testing :)

Reply:
v1.0.0 released last Friday, just been cleaning up FAQ etc